utils/pmx/pmx.1
.\" This manpage is licensed under the GNU Public License
.TH pmx 1 2011-11-28  "pmx script, version 0.2" ""

.SH NAME
pmx \- This script processes pmx files and then deletes intermediate files

.SH SYNOPSIS
[ texlua ]
.B pmx[.lua]
{
.IR option
|
.IR jobname 
[ .pmx ] 
} ...
.TP 
The default effect is to process each file using 
.TP 12
.BR pmxab (1)
(pre-processing pass)
.TP 
.BR etex (1)
(first pass)
.TP
.BR musixflx (1)
(second pass)
.TP
.BR etex (1)
(third pass)
.TP
.BR dvips (1)
(dvi to Postscript)
.TP
.BR ps2pdf (1)
(Postscript to Portable Document Format)
.TP
followed by removal of intermediate files.

.SH OPTIONS
.TP 8
.B -v 
output program name and version number, and quit
.TP
.B -h
output usage summary and quit
.TP
.B -l 
changes the TeX processor to
.BR latex (1)
or
.BR pdflatex (1)
.TP
.B -p 
changes the TeX processor to 
.BR pdfetex (1) 
or
.BR pdflatex (1)
(and doesn't run 
.BR dvips (1)
and 
.BR ps2pdf (1)
or
.BR dvipdfm (1)
)
.TP
.B -d 
changes the dvi processor to
.BR dvipdfm (1)
.TP
.B -s 
stops processing at the dvi file
.TP
.B -t
stops processing at the tex (and midi) files
.TP 
.B -i
retain intermediate files
.TP
.B -f
default processing 
.TP 0
Options affect processing of all succeeding PMX files unless changed by another option.

.SH SEE ALSO
.BR pmxab (1)
.BR scor2prt (1)
.BR etex (1),
.BR latex (1),
.BR pdfetex (1),
.BR pdflatex (1),
.BR musixflx (1),
.BR dvipdfm (1),
.BR dvips (1),
.BR ps2pdf (1).
.br
pmx260.pdf
.br
pmxccn.pdf

.SH AUTHOR 
This manual page was written by Bob Tennent <rdt@cs.queensu.ca>.


ChangeLog
2012-04-03  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2012-04-03.

2012-03-31  Karl Berry  <karl@tug.org>

	* Build (TL_CONF_LARGEFILE): switch default to --enable-largefile.
	More info in README.config.

2012-03-12  Karl Berry  <karl@tug.org>

	* README.config: 32-bit rather than 32Bit

2012-03-11  Karl Berry  <karl@tug.org>

	* README.config: text tweaks for poppler/libxpdf/luatex/etc.

2012-01-16  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* configure.ac: Report `$enable_largefile'.

2012-01-13  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2012-01-13.

2011-12-22  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am: Replace DISTCHECK_CONFIGURE_FLAGS by
	AM_DISTCHECK_CONFIGURE_FLAGS (for automake-1.11.2).

2011-12-06  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* configure.ac: Use TL specific installation directories only
	for $enable_native_texlive_build=yes (also datarootdir).
	Display the installation directories.

	* README.config, README.distro: Adapted.

2011-10-05  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2011-10-05.

2011-08-10  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2011-08-10.

2011-07-12  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am [EXTRA_DIST]: Add m4/ChangeLog.

2011-07-04  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* version.ac: Switch 2011 => 2012/dev.

2011-06-21  Karl Berry  <karl@tug.org>

	* Build: report final exit status.

2011-04-15  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2011-06-15.

2011-04-14  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am, configure.ac: Conditional 'cross_compiling' =>
	'cross' as in libs/icu, texk/dvipng, and texk/web2c.

	* version.ac: Switch 2011/dev => 2011.

2011-04-12  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2011-04-12.

2011-04-09  Karl Berry  <karl@tug.org>

	* Build (MACOSX_DEPLOYMENT_TARGET): no longer needed per koch.

2011-03-03  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2011-03-03.

2011-02-16  Karl Berry  <karl@tug.org>

	* Build: include uname -a in log, mention log if failure.

2011-01-19  Karl Berry  <karl@tug.org>

	* Build: don't try to count binaries if no bindir.	

2010-10-24  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac (new): Define the current TeX Live tarball version.
	* version.ac: Switch to 2011/dev.

2010-07-06  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* version.ac (new): Define the current TeX Live version.

2010-05-18  Karl Berry  <karl@tug.org>

	* Build (TL_CONF_BANNER): switch ident to /pretest.

2010-03-24  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Build (TL_CONF_THREADS): Removed, now in libs/icu/configure.

2010-03-05  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am (triptrap): Run intermediate triptrap target in
	 texk/, texk/web2c/Makefile might not yet exist.

2010-03-04  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am (triptrap): New target to run TRIP and TRAP tests.
	Do not automatically run 'make check' when cross compiling.

2010-03-03  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am, configure.ac: Do not automatically run tests as
	part of 'make world' when cross compiling.

2010-02-22  Karl Berry  <karl@tug.org>

	* Build (TL_CONF_BANNER): 2010/dev.

2009-12-24  Karl Berry  <karl@tug.org>

	* Makefile.am (world): also run make check.
	(check_target): new variable.
	(skip): new target, to do nothing.

	* Build (TL_CONF_BANNER): add /dev.

2009-12-22  Karl Berry  <karl@tug.org>

	* build-aux/ylwrap: update from current automake (no real changes).

2009-11-12  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* configure.ac: Disable cross compilation of xindy.
	* configure.ac: Report absolute build and source dir; reject
	paths with embedded spaces.

2009-09-17  Karl Berry  <karl@tug.org>

	* Build: show Build command itself in build.log.

2009-09-01  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* configure.ac: Report level of compiler warnings used.

2009-08-19  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am (dist*check_listfiles): Stricter checking for
	'make distcheck': check for leftover files and symlinks.
	This really ought to be the Automake default.

2009-06-18  Karl Berry  <karl@tug.org>

	* Build: eradicate -Wdecl stuff, peb has implemented something
	more general.

2009-06-01  Karl Berry  <karl@tug.org>

	* Makefile.am (world): don't echo the echo.

2009-05-22  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am (world): serialize to avoid parallel build problems.

2009-05-21  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Build (TL_CONF_MAC_FONTS, TL_CONF_XINDY): removed,
	no longer required.

2009-05-18  Karl Berry  <karl@tug.org>

	* Build (-Wdecl): allow with -g, to do -Wdeclaration-after-statement.

2009-05-16  Karl Berry  <karl@tug.org>

	* Build (TL_CONF_XINDY): default to disabling latex-requiring stuff.

2009-04-29  Karl Berry  <karl@tug.org>

	* Build (MACOSX_DEPLOYMENT_TARGET): set to 10.3 on powerpc-darwin
	(but not i386-darwin).

2009-04-11  Karl Berry  <karl@tug.org>

	* Makefile.am (world): do the run-texlinks target in texk/tetex
	to make the per-format links to binaries.

2009-03-14  Karl Berry  <karl@tug.org>

	* Build: include timestamps in output.
	(TL_BUILD_ENV): include OBJCFLAGS for xdv2pdf.

2009-03-13  Karl Berry  <karl@tug.org>

	* Build (bindir): call them executables, not binaries.
	Use set -vx, not +vx.

2009-03-08  Karl Berry  <karl@tug.org>

	* Build (TL_CONF_SHARED): new variable, --disable-shared by default.
	(-g): alias for --debug.
	use set -vx to see configure && make commands in the log.	

2009-01-24  Karl Berry  <karl@tug.org>

	* utils/texinfo: remove, we will no longer include Texinfo in TeX Live
	(except for texinfo.tex).
	In practice it is more easily available separately,
	and the (n)curses build dependency is a useless pain.
	* utils/README,
	* configure.in,
	* Makefile.in,
	* README,
	* tetex.ac,

utils/pmx/pmx.lua
#!/usr/bin/env texlua  

VERSION = "0.2"

--[[
     pmx.lua: processes MusiXTeX files using pmxab as a pre-processor (and deletes intermediate files)

     (c) Copyright 2011 Bob Tennent rdt@cs.queensu.ca

     This program is free software; you can redistribute it and/or modify it
     under the terms of the GNU General Public License as published by the
     Free Software Foundation; either version 2 of the License, or (at your
     option) any later version.

     This program is distributed in the hope that it will be useful,
     but WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
     Public License for more details.

     You should have received a copy of the GNU General Public License along
     with this program; if not, write to the Free Software Foundation, Inc.,
     51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.

--]]

--[[

  ChangeLog:
     version 0.2  2011-11-28 RDT
       added -i (retain intermediate files) option

     version 0.1  2011-07-12 RDT

--]]

function usage()
  print("Usage:  [texlua] pmx.lua { option | basename[.pmx] } ... ")
  print("options: -v  version")
  print("         -h  help")
  print("         -l  latex (or pdflatex)")
  print("         -p  pdfetex (or pdflatex)")
  print("         -d  dvipdfm")
  print("         -s  stop at dvi")
  print("         -t  stop at tex/mid")
  print("         -i  retain intermediate files")
  print("         -f  restore default processing")
end

function whoami ()
  print("This is pmx.lua version ".. VERSION .. ".")
end

whoami()
if #arg == 0 then
  usage()
  os.exit(0)
end

-- defaults:
tex = "etex"  
musixflx = "musixflx"
dvi = "dvips"
ps2pdf = "ps2pdf"
intermediate = 1

exit_code = 0
narg = 1
repeat
  this_arg = arg[narg]
  if this_arg == "-v" then
    os.exit(0)
  elseif this_arg == "-h" then
    usage()
    os.exit(0)
  elseif this_arg == "-l" then
    if tex == "pdfetex" then
      tex = "pdflatex"
    else
      tex = "latex"
    end
  elseif this_arg == "-p" then
    if tex == "latex" then
      tex = "pdflatex"
    else
      tex = "pdfetex"
    end
    dvi = ""; ps2pdf = ""
  elseif this_arg == "-d" then
    dvi = "dvipdfm"; ps2pdf = ""
  elseif this_arg == "-s" then
    dvi = ""; ps2pdf = ""
  elseif this_arg == "-f" then
    tex = "etex"; dvi = "dvips"; ps2pdf = "ps2pdf"; intermediate = 1
  elseif this_arg == "-t" then
    tex = ""; dvi = ""; ps2pdf = ""
  elseif this_arg == "-i" then
    intermediate = 0
  else
    filename = this_arg 
    if filename ~= "" and string.sub(filename, -4, -1) == ".pmx" then
        filename = string.sub(filename, 1, -5)
    end
    if not io.open(filename .. ".pmx", "r") then
      print("Non-existent file: ", filename .. ".pmx")
    else
      print("Processing ".. filename .. ".pmx.")
      os.remove( filename .. ".mx2" )
      os.execute("pmxab" .. " " .. filename )
      pmxaerr = io.open("pmxaerr.dat", "r")
      if (not pmxaerr) then
        print("No log file.")
        os.exit(1)
      end
      linebuf = pmxaerr:read()
      err = tonumber(linebuf)
      pmxaerr:close()
      if ( err == 0 ) and
         ((tex == "") or os.execute(tex .. " " .. filename) == 0) and
         ((tex == "") or os.execute(musixflx .. " " .. filename) == 0) and
         ((tex == "") or os.execute(tex .. " " .. filename) == 0) and
         ((tex ~= "latex" and tex ~= "pdflatex") 
           or (os.execute(tex .. " " .. filename) == 0)) and
         (dvi == "" or  (os.execute(dvi .. " " .. filename) == 0)) and
         (ps2pdf == "" or (os.execute(ps2pdf .. " " .. filename .. ".ps") == 0) )
      then 
        if ps2pdf ~= "" then
          print(filename .. ".pdf generated by " .. ps2pdf .. ".")
        end
        if intermediate == 1 then -- clean-up:
          os.remove( "pmxaerr.dat" )
          os.remove( filename .. ".mx1" )
          os.remove( filename .. ".mx2" )
          if dvi ~= "" then
            os.remove( filename .. ".dvi" )
          end
          if ps2pdf ~= "" then 
            os.remove( filename .. ".ps" )
          end
        end
      else
        print("pmx/MusiXTeX processing of " .. filename .. ".pmx fails.\n")
        exit_code = 2
        --[[ uncomment for debugging
        print("tex = ", tex)
        print("dvi = ", dvi)
        print("ps2pdf = ", ps2pdf)
        --]]
      end

    end --if not io.open ...
  end --if this_arg == ...
  narg = narg+1
until narg > #arg 
os.exit( exit_code )
